### PR TITLE
TS number tests with embedded study

### PR DESCRIPTION
**Pull request by guilpier-code**

### CODE REVIEW

File: src/tests/src/solver/simulation/tests-ts-numbers.cpp [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+void addClusterToAreaList(Area* area, std::shared_ptr<ClusterType> cluster)
+{}
+
+template<>
+void addClusterToAreaList(Area* area, std::shared_ptr<ThermalCluster> cluster)
+{
+	area->thermal.clusters.push_back(cluster.get());
+	area->thermal.list.add(cluster);
+	area->thermal.list.mapping[cluster->id()] = cluster;
+}
```
Comment (flomnes) [inline src/tests/src/solver/simulation/tests-ts-numbers.cpp:55]: If the thermal cluster is "must-run", it should be added to mustRunList instead

Comment (flomnes) [reply src/tests/src/solver/simulation/tests-ts-numbers.cpp:55]: ~~l. 50 : `area->thermal.clusters` is of type `std::map<string, std::shared_ptr<ThermalCluster>>`. Therefore there is no need to call `get()` to access the raw pointer.~~

~~Doing so may lead to a double free (happened in the past).~~

I was wrong.

File: src/tests/src/solver/simulation/tests-ts-numbers.cpp [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+	// ... Area 2 : thermal cluster 1
+	auto thCluster_21 = addClusterToArea<ThermalCluster>(area_2, "th-cluster-21");
+	thCluster_21->series->series.resize(4, 1);
+
+	area_2->thermal.resizeAllTimeseriesNumbers(1);
+
+	BOOST_CHECK(Generate(*study));
+}
+
+BOOST_AUTO_TEST_CASE(two_areas_3_thermal_clusters_with_different_number_of_ready_made_ts___check_intra_modal_ok)
```
Comment (flomnes) [inline src/tests/src/solver/simulation/tests-ts-numbers.cpp]: Should be **ko**

**Review by flomnes — COMMENTED** [review]
I mostly agree with this way to do things.

Add intermodal_ko tests ?

File: src/tests/src/solver/simulation/tests-ts-numbers.cpp [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+	area_1->resizeAllTimeseriesNumbers(1 + study->runtime->rangeLimits.year[rangeEnd]);
+	area_1->load.series->series.resize(5, 1);
+
+	// Area 2
+	Area* area_2 = addAreaToStudy(study, "Area 2");
+	area_2->resizeAllTimeseriesNumbers(1 + study->runtime->rangeLimits.year[rangeEnd]);
+	area_2->load.series->series.resize(5, 1);
+
+	BOOST_CHECK(Generate(*study));
+}
```
Comment (flomnes) [inline src/tests/src/solver/simulation/tests-ts-numbers.cpp]: Check that the TS numbers satisfy the intra-modal rule here.

File: src/tests/src/solver/simulation/tests-ts-numbers.cpp [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+	auto thCluster_21 = addClusterToArea<ThermalCluster>(area_2, "th-cluster-21");
+	thCluster_21->series->series.resize(4, 1);
+
+	area_2->resizeAllTimeseriesNumbers(1 + study->runtime->rangeLimits.year[rangeEnd]);
+
+	BOOST_CHECK(Generate(*study));
+}
+
+BOOST_AUTO_TEST_CASE(two_areas_3_thermal_clusters_with_different_number_of_ready_made_ts___check_intra_modal_OK)
+{
```
Comment (flomnes) [inline src/tests/src/solver/simulation/tests-ts-numbers.cpp]: KO

File: src/tests/src/solver/simulation/tests-ts-numbers.cpp (hunk 3 above)
Comment (guilpier-code) [reply src/tests/src/solver/simulation/tests-ts-numbers.cpp]: done

File: src/tests/src/solver/simulation/tests-ts-numbers.cpp (hunk 4 above)
Comment (guilpier-code) [reply src/tests/src/solver/simulation/tests-ts-numbers.cpp]: done

File: src/tests/src/solver/simulation/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+	${src_solver_simulation}/timeseries-numbers.cpp
+)
+	
+add_library(lib-needed-by-ts-numbers ${SRC_TS_NUMBERS_LIB})
+
+target_link_libraries(lib-needed-by-ts-numbers
+						PUBLIC
+							libantares-core
+							libmodel_antares
+)
```
Comment (flomnes) [inline src/tests/src/solver/simulation/CMakeLists.txt]: Why create a lib ? We could just add `libantares-core` and `libmodel_antares` after line 41 instead.

File: src/tests/src/solver/simulation/tests-ts-numbers.cpp [hunk 6]
```diff
@@ -0,0 +1,518 @@
...
+	uint windTsNumber = area->wind.series->timeseriesNumbers[0][year];
+	uint solarTsNumber = area->solar.series->timeseriesNumbers[0][year];
+	uint hydroTsNumber = area->hydro.series->timeseriesNumbers[0][year];
+	uint thermalTsNumber = thCluster->series->timeseriesNumbers[0][year];
+
+	BOOST_TEST((loadTsNumber >= 0 && loadTsNumber < loadNumberOfTs));
+	BOOST_TEST((windTsNumber >= 0 && windTsNumber < windNumberOfTs));
+	BOOST_TEST((solarTsNumber>= 0 && solarTsNumber < solarNumberOfTs));
+	BOOST_TEST((hydroTsNumber >= 0 && hydroTsNumber < hydroNumberOfTs));
+	BOOST_TEST((thermalTsNumber>= 0 && thermalTsNumber < thermalNumberOfTs));
```
Comment (flomnes) [inline src/tests/src/solver/simulation/tests-ts-numbers.cpp]: Remove useless parentheses.

**Review by flomnes — CHANGES_REQUESTED** [review]

File: src/solver/simulation/timeseries-numbers.cpp [hunk 7]
```diff
@@ -63,19 +54,22 @@ const map<TimeSeries, string> ts_to_tsTitle = {
...
-        {                            \
-            if (first)               \
-            {                        \
-                e += ", ";           \
-                first = false;       \
-            }                        \
-            e += TEXT;               \
-        }                            \
-    } while (0)
+void addInterModalTimeSeriesToMessage(const bool* isTSintermodal, std::string & interModalTsMsg)
```
Comment (flomnes) [inline src/solver/simulation/timeseries-numbers.cpp]: `std::array<bool, size>`

File: src/solver/simulation/timeseries-numbers.cpp [hunk 8]
```diff
@@ -689,22 +666,19 @@ bool TimeSeriesNumbers::Generate(Data::Study& study)
...
+    uint nbTimeseriesByMode[timeSeriesCount];
 
-    uint32 intramodal_draws[Data::timeSeriesCount] = {0, 0, 0, 0, 0, 0};
+    uint32 intramodal_draws[timeSeriesCount] = {0, 0, 0, 0, 0, 0};
 
-    const bool isTSgenerated[Data::timeSeriesCount]
+    const bool isTSgenerated[timeSeriesCount]
       = {0 != (Data::timeSeriesLoad & parameters.timeSeriesToRefresh),
          0 != (Data::timeSeriesHydro & parameters.timeSeriesToRefresh),
          0 != (Data::timeSeriesWind & parameters.timeSeriesToRefresh),
```
Comment (flomnes) [inline src/solver/simulation/timeseries-numbers.cpp]: Use `std::array` for `nbTimeseriesByMode`, `intramodal_draws` and `isTSgenerated` as well ?

File: src/tests/src/solver/simulation/CMakeLists.txt (hunk 5 above)
Comment (guilpier-code) [reply src/tests/src/solver/simulation/CMakeLists.txt]: I wanted to separate sources from code and tests sources. 
Maybe not relevant.
Changes were done here.

File: src/tests/src/solver/simulation/tests-ts-numbers.cpp (hunk 6 above)
Comment (guilpier-code) [reply src/tests/src/solver/simulation/tests-ts-numbers.cpp]: done

File: src/solver/simulation/timeseries-numbers.cpp (hunk 7 above)
Comment (guilpier-code) [reply src/solver/simulation/timeseries-numbers.cpp]: done

File: src/solver/simulation/timeseries-numbers.cpp (hunk 8 above)
Comment (guilpier-code) [reply src/solver/simulation/timeseries-numbers.cpp]: done

**Pull request merged**